pandas/core/reshape/reshape.py

```diff
@@ -779,7 +778,22 @@ def melt(frame, id_vars=None, value_vars=None, var_name=None,
         mdata[col] = np.asanyarray(frame.columns
                                    ._get_level_values(i)).repeat(N)
 
-    return DataFrame(mdata, columns=mcolumns)
+    result = DataFrame(mdata, columns=mcolumns)
+
+    if keep_index:
+        orig_index_values = list(np.tile(frame.index.get_values(), K))
+
```

Comment:
this is quite awkward, you have several cases which you need to disambiguate. e.g. if the original is a MI or not.

Reply:
Thanks @jreback for looking over my code and the comment.

I think what I wrote should work with any number of levels.

E. g.
```python
arrays = [['bar', 'bar', 'baz', 'baz', 'foo', 'foo', 'qux', 'qux'],
          ['one', 'two', 'one', 'two', 'one', 'two', 'one', 'two']]
tuples = list(zip(*arrays))
idx_multi = pd.MultiIndex.from_tuples(tuples)
idx_single = pd.Index(arrays[0])

# Index
print(list(np.tile(idx_single, 1)))
print(list(np.tile(idx_single, 2)))

# MultiIndex
print(list(np.tile(idx_multi, 1)))
print(list(np.tile(idx_multi, 2)))
```
But do I have to make it more explicit (= Pythonic)? Or did I miss something else?